module/zfs/dsl_dataset.c

```diff
@@ -599,21 +599,23 @@ dsl_dataset_hold_obj_flags(dsl_pool_t *dp, uint64_t dsobj,
 			}
 		}
 	}
-	ASSERT3P(ds->ds_dbuf, ==, dbuf);
-	ASSERT3P(dsl_dataset_phys(ds), ==, dbuf->db_data);
-	ASSERT(dsl_dataset_phys(ds)->ds_prev_snap_obj != 0 ||
-	    spa_version(dp->dp_spa) < SPA_VERSION_ORIGIN ||
-	    dp->dp_origin_snap == NULL || ds == dp->dp_origin_snap);
-	*dsp = ds;
 
 	if ((flags & DS_HOLD_FLAG_DECRYPT) && ds->ds_dir->dd_crypto_obj != 0) {
-		err = spa_keystore_create_mapping(dp->dp_spa, ds, ds);
+		err = spa_keystore_create_mapping(dp->dp_spa, ds, ds,
+		    &ds->ds_key_mapping);
```

Comment:
This adds a hold on the mapping, right?  When is that released?

Reply:
`dsl_dataset_rele_flags()`